go.mod

```diff
@@ -22,8 +22,6 @@ replace (
 	k8s.io/cluster-bootstrap => k8s.io/cluster-bootstrap v0.16.4
 	k8s.io/code-generator => k8s.io/code-generator v0.18.3
 	k8s.io/kube-openapi => k8s.io/kube-openapi v0.0.0-20191107075043-30be4d16710a
-	kubevirt.io/client-go => github.com/rancher/kubevirt-client-go v0.20.2-0.20201124032527-ff3943be35bf
-	kubevirt.io/containerized-data-importer => github.com/rancher/kubevirt-containerized-data-importer v1.22.1-0.20201127032458-4f73ed532d31
```

Comment:
those replace are required for our framework specification, u should be able to found the correct commit from https://github.com/rancher/kubevirt-client-go/tree/v0.37.2-rancher and https://github.com/rancher/kubevirt-containerized-data-importer/tree/v1.28.3-rancher

Reply:
Hi @guangbochen , thank you for review. Actually I was trying to understand if its possible to somehow avoid those replaces. IMHO having to maintain a fork of `kubevirt/client-go` with patches adds complexity and requires additional effort (especially in long-term perspective). Maybe it is worth trying to make it simpler, what do you think?

I looked at the patches which are applied to the `kubevirt/client-go`. From what I can see they basically do two things:

1. Modify the original `client-go` sources (by adding `genclient`, `deepcopy` and `SchemeGroupVersion`)
2. Copy `v1` API to `v1alpha3` (thus creating a new go package with `v1alpha3` APIs identical to `v1`)

As far as I understand the changes in **1)** are needed for harvester. But I am not sure about **2)**. If we update all the refences in the harvester source code pointing to `v1` API instead of `v1alpha3` that should make **2)** redundant, right? I already tried to do that within this PR. It should be pretty safe since the APIs are the same. Then regarding **1)**, what do you think if we try to upstream those changes? E.g. I could create a PR in `kubevirt` and add those to `client-go`. It will not affect the situation this time but will likely make the next version bump easier (well, if the changes are accepted though...).

I hope that my understanding of the current state is correct and I havent missed any details. So what do you think about it in general? I.e. switching to `v1` in harvester sources, trying to upstream the changes and to eliminate the need for the fork? Would be great to know your opinion.

Reply:
yes, this is something we could make an enhancement so the Harvester can use the dependency libraries directly instead of doing the folk and modification, would be appreciated if u can help to managed to submit those changes to the upstream, thanks.

Reply:
Great, then I will try to submit the changes. Lets see if they are accepted. I will mention you in the PR so you can also track the status.

Reply:
BTW, small question:

The patch https://github.com/rancher/kubevirt-client-go/commit/ad1c81522f314edfd334225bffde41659ffd52cb also adds `deepcopy` for `VirtualMachineInstanceGuestOSUser`:

```go
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
type VirtualMachineInstanceGuestOSUser struct {
```

I was not able to find any direct reference to `VirtualMachineInstanceGuestOSUser` in Harvester code. Maybe it is not needed anymore?

Reply:
Yep, I think the `VirtualMachineInstanceGuestOSUser ` is currently not used by the Harvester, but would prefer to make the change to be consistent for future usage.

Reply:
Hm... in fact it causes troubles in kubevirt:

```
../../vendor/kubevirt.io/client-go/api/v1/deepcopy_generated.go:2927:3: cannot use c (type *VirtualMachineInstanceGuestOSUser) as type "k8s.io/apimachinery/pkg/runtime".Object in return argument:
	*VirtualMachineInstanceGuestOSUser does not implement "k8s.io/apimachinery/pkg/runtime".Object (missing GetObjectKind method)
```

Reply:
I think we can skip it for now since we are not using it or anytime soon in the future based on my expectation.

Reply:
Okay, sounds good :+1:  We can address this separately when it becomes relevant.